Annotate node in schema to allow any type

gcheadle-vmware opened the pull request.

DennisDenuto committed:
Add @schema/type any annotation for arrays and maps

Signed-off-by: Garrett Cheadle <gcheadle@vmware.com>

DennisDenuto committed:
Update go version in go.mod

gcheadle-vmware committed:
Add error handling for schema/type

gcheadle-vmware committed:
Change AnySchema to DocumentSchema with AnyType Value

Signed-off-by: Garrett Cheadle <gcheadle@vmware.com>

cari-lynn commented on pkg/cmd/template/schema_test.go:
```diff
@@ -646,6 +646,198 @@ overriden:
+		})
+
+		assertSucceeds(t, filesToProcess, expected, opts)
+	})
```
I believe that this case where data values aren't given is the same as this test: `TestSchema_Provides_default_values`. So I think we can remove this test.

cari-lynn commented on pkg/cmd/template/schema_test.go:
```diff
@@ -646,6 +646,198 @@ overriden:
+
+		assertFails(t, filesToProcess, expectedErr, opts)
+	})
+}
```
This tests may fit under `TestSchema_When_invalid_reports_error` since that's a kind of catch-all for schema invalid errors.

cari-lynn commented on pkg/cmd/template/schema_test.go:
```diff
@@ -646,6 +646,198 @@ overriden:
+		})
+
+		assertSucceeds(t, filesToProcess, expected, opts)
+	})
```
I think narrowing the scope of this test to only test map items, and creating a separate test to test array items would fit with the pattern introduced with `TestSchema_Allows_null_values_via_nullable_annotation`. It would also make this test simpler and more focused which may make the tests easier to modify and add to in the future.

cari-lynn commented on pkg/cmd/template/schema_test.go:
```diff
@@ -646,6 +646,198 @@ overriden:
+`
+		dataValuesYAML := `#@data/values
+---
+#@overlay/replace
```
I think I'm missing something here, I don't understand why the `#@overlay/replace` is necessary. Shouldn't this test work without it?

gcheadle-vmware replied:
This relates to #344. The overlay/replace allows us to get past the overlaying of data values without erroring. Without the overlay/replace, we get overlay errors about expecting a string but found a map. In this test, the string is the schema value of a node that is annotated as anyType, thus replacing the string with any value should be allowed by the schema (which it is).

gcheadle-vmware replied:
I was able to remove the annotation from the array item, and it did not error. So I'm going to look into this more to see if there is a way to make it possible for maps.

gcheadle-vmware replied:
The array item does not need the `#@overlay/replace` due to the schema's default data values. When we construct the default dvs for an array, we set the items to be empty: `typedNode.Items = []*yamlmeta.ArrayItem{}`. Since we can overlay onto `nil`, it overlays without the replace annotation. However, arrays are a special case in schemas since they do not have default values.

cari-lynn replied:
That makes sense why it doesn't work now. IIRC, this is the only place where we require an overlay annotation on the first data values document, so it's a divergent experience. Curious if you have thoughts on whether it would be possible to tackle here or if it makes sense to separate out?

gcheadle-vmware replied:
Since #344 has been unprioritized, John and I discussed where this will be addressed in our future work. The conclusion that we came to was that we will not need the overlay annotations once the data values are passed using the `--data-values-file` flag. When the `--data-values-file` feature is implemented, this test, along with ['when map item's key is not among those declared in schema'](https://github.com/vmware-tanzu/carvel-ytt/blob/91cd2f10285ee2a456005bd5a3d9fad3cdf14baa/pkg/cmd/template/schema_test.go#L183), and ['checks after every data values document is processed and stops if there was a violation'](https://github.com/vmware-tanzu/carvel-ytt/blob/91cd2f10285ee2a456005bd5a3d9fad3cdf14baa/pkg/cmd/template/schema_test.go#L380) should pass without any overlay annotations.

cari-lynn commented on pkg/schema/schema.go:
```diff
@@ -96,6 +97,13 @@ func NewMapItemType(item *yamlmeta.MapItem) (*MapItemType, error) {
+	if err != nil {
+		return nil, NewInvalidSchemaTypeAnnotationError(item, err.Error())
+	} else if anyType {
+		valueType = AnyType{Position: item.Position}
```
The `else` can be removed here since the previous if returns and we're guaranteed to be in the 'else' case. (The same goes for the code in `NewArrayItemType` line 148).

cari-lynn commented on pkg/schema/schema.go:
```diff
@@ -167,6 +181,30 @@ func newCollectionItemValueType(collectionItemValue interface{}, position *filep
 
+func processTypeAnnotation(templateAnnotations template.NodeAnnotations) (bool, error) {
+	if ann, anyType := templateAnnotations[AnnotationSchemaType]; anyType {
+		kwargName, err := core.NewStarlarkValue(ann.Kwargs[0][0]).AsString()
```
We should do some validation that `ann.Kwargs[0][0]` exists. If someone accidentally types `#@schema/type any` instead of `#@schema/type any=True` this panics.

gcheadle-vmware replied:
I found that that panic when providing `#@schema/type any` comes from `ann.Kwargs[0]` being out of range, thus I think it is sufficient enough to check the length of `ann.Kwargs`.

However, one thing that I discovered that I could not figure out how to avoid  were these scenerios: 
- `#@schema/type any=something` gives this unhelpful error:
```
- undefined: something
    schema.yml:3 | #@schema/type any=something
````
however, we get an actual schema error message is `something` is in quotes.
- `#@schema/type any True` gives this unhelpful error:
```
- got identifier, want ','
    schema.yml:3 | #@schema/type any True
```
which seem to error during the EvalYaml call.
It doesn't feel like this issue is in scope due to the fact that I have seen this error before when using other annotations.

cari-lynn replied:
Yeah I agree, that seems out of scope since those are similar errors to what we see in other areas of code now. 👍

cari-lynn commented on pkg/schema/schema.go:
```diff
@@ -167,6 +181,30 @@ func newCollectionItemValueType(collectionItemValue interface{}, position *filep
+		if err != nil {
+			return false, err
+		}
+		if kwargName != "any" {
```
"any" could be a constant.

cari-lynn commented on pkg/schema/schema.go:
```diff
@@ -167,6 +181,30 @@ func newCollectionItemValueType(collectionItemValue interface{}, position *filep
+		}
+		kwargBool, err := core.NewStarlarkValue(ann.Kwargs[0][1]).AsBool()
+		if kwargBool == false && err != nil {
+			kwargBool := core.NewStarlarkValue(ann.Kwargs[0][1]).AsGoValue()
```
As a stylistic thing I would probably assign this to a new variable rather than reassign it to `kwargBool ` since they are different types and `kwargBool` is used later on as well.

pivotaljohn replied:
+1.

While we're talkin' names, here. At this point, I think we're certain of the _role_ of the second kwarg: it's whether `any` type is allowed. Surfacing the role into the name (a form of "revealing intention") can make the code more readable. 👍🏼

cari-lynn commented on pkg/schema/schema.go:
```diff
@@ -167,6 +181,30 @@ func newCollectionItemValueType(collectionItemValue interface{}, position *filep
 	return nil, fmt.Errorf("Collection item type did not match any known types")
 }
 
+func processTypeAnnotation(templateAnnotations template.NodeAnnotations) (bool, error) {
```
Since this function returns a bool, maybe rename it to indicate that. `hasAnyTypeAnnotation()`?

pivotaljohn commented on pkg/schema/schema.go:
```diff
@@ -129,11 +137,17 @@ func NewArrayItemType(item *yamlmeta.ArrayItem) (*ArrayItemType, error) {
+	if _, found := templateAnnotations[AnnotationSchemaNullable]; found {
 		return nil, NewInvalidSchemaError(item, fmt.Sprintf("@%s is not supported on array items", AnnotationSchemaNullable), "")
 	}
+	anyType, err := processTypeAnnotation(templateAnnotations)
```
Can we consolidate the ways we check for the "various" (😁  there's two right now) annotations? Not sure if it fits, but I have an impulse to use a `switch` here, for example.

gcheadle-vmware replied:
with `schema/nullable` and `schema/type`, is it possible to annotate a node with both annotations? If not, I agree that a switch statement would be the way to go.

gcheadle-vmware replied:
is this a valid schema definition?
```
#@schema/match data_values=True
---
#@schema/type any=True
foo: null
```
just as this is a valid schama:
```
#@schema/match data_values=True
---
#@schema/nullable
foo: null
```

cari-lynn replied:
This may have already been resolved offline, but my take here is yes this is valid

pivotaljohn replied:
+1 @cari-lynn.

We _did_ talk about it offline. Meant to summarize here. Will do now.

That:
- "null" is the one value of the type "null" and so it is a legitimate value to have in a schema
- `type any=True` implies _all_ types, which includes "null" type

And:
- while `@schema/nullable` over a "null" is technically valid, it's also useless (a variable that can't vary?) but not seeing a need to attempt to prevent it, right now.
- one use-case of an author using `type any=True` is a required key whose contents are merely passed along: "null" seems like a reasonable default, here.

Did I miss anything @gcheadle-vmware ?

gcheadle-vmware committed:
Address feedback on tests

gcheadle-vmware committed:
Rename variables and add error case

- Error case for when schema/type annotation is incomplete

gcheadle-vmware committed:
Process mutally exclusive schema annoations with switch statement

gcheadle-vmware committed:
Error when both schema/type and schema/nullable

- update invalid schema annoation error message

pivotaljohn commented on pkg/schema/schema.go:
```diff
@@ -88,14 +90,22 @@ func NewMapItemType(item *yamlmeta.MapItem) (*MapItemType, error) {
-	if _, nullable := templateAnnotations[AnnotationSchemaNullable]; nullable {
-		defaultValue = nil
-	} else if valueType == nil {
+	templateAnnotations, ann, err := processSchemaNodeAnnotations(item)
```
I'm finding it challenging to extract the general algorithm to deduce the type of an item.
All the ingredients are here... but they are squeezed together... this is making it hard for me to tease apart the various responsibilities, which is how I try to make sense of a chunk of code.
I think we should attempt to create some space between the various responsibilities.

### Taking Inventory of Responsibilities
What are these "responsibilities"?

Starting at `NewMapItemType()`, I see these responsibilities:
- extracting values from the kwargs of annotations (including converting from Starlark to Go)
- validating those values
- determining whether two types can co-exist (said another way, whether they can combine into a single type)
- "calculating" the default value of the type

How might we figure out how to "better" organize these?

### Sketching

I did two things:
1. wrote pseudo code of an "algorithm" — the minimum set of steps, involving the smallest number of things.
2. drew a picture of what could be constituent parts — what are all the things? what categories do they fall into? what are the responsibilities of each category?

#### Pseudo Code
Here's that pseudo code:

```
determine the type:
  get the inferred type from the item's value. 
  get the explicit type from the item's annotations.
  if neither yielded a valid type  (e.g. _only_ "null" type)
    err
  if the explicit type can combine the inferred type,
    type = (explicit or inferred)
  else
    type = (explicit)

determine the default value:
  get the inferred default value from the item's value
  get the explicit default value from the item's annotations
  if explicit value exists
    default = explicit
  else
    default = implicit

return &item{valueType: type, defaultValue: default}
```

#### A "Class" Diagram

And I used Miro for that picture: [(Miro Board) "ytt boxes and lines : Elements of Schema"](https://miro.com/app/board/o9J_kopV1_U=/?moveToWidget=3074457358481979842&cot=14)
_(It's way more than the scope of what we're looking at, but I was trying to get perspective, so I zoomed out a bit — thinking about the bigger picture.)_

FWIW, it felt like the pseudo code came out of "top-down" thinking and the diagram came out of "bottom-up" thinking.

### Reimagining `NewMapItemType()`

With these in hand, what if we reimagined `NewMapItemType()` so that it:
- delegates:
  - getting the `Type` from any annotations on the `MapItem` (and all that's involved)
  - inferring the `Type` from the value of the `MapItem`
  - getting the default value from the `MapItem`
- takes responsibility for:
  - given the inferred and explicit `Type`s determine the final `Type`
  - given final Type, default Value, etc... build the `MapItemType`

Thoughts?

pivotaljohn commented on pkg/schema/type.go:
```diff
@@ -338,6 +360,39 @@ func (m *ScalarType) AssignTypeTo(typeable yamlmeta.Typeable) (chk yamlmeta.Type
 	panic(fmt.Sprintf("Attempt to assign a type to a scalar. (scalars are not nodes in the AST) m=%#v ; typeable=%#v", m, typeable))
 }
 
+func (a AnyType) AssignTypeTo(typeable yamlmeta.Typeable) (chk yamlmeta.TypeCheck) {
+
+	switch typedItem := typeable.(type) {
+	case *yamlmeta.Array:
+		typeable.SetType(&ArrayType{ItemsType: &ArrayItemType{ValueType: a, Position: a.Position}})
+		for _, arrayItem := range typedItem.Items {
+			a.AssignTypeTo(arrayItem)
+		}
+	case *yamlmeta.Map:
+		var mItemTypeS []*MapItemType
+		for _, mapItem := range typedItem.Items {
+			a.AssignTypeTo(mapItem)
+			mItemType := &MapItemType{Key: mapItem.Key, ValueType: a, Position: a.Position}
+			mItemTypeS = append(mItemTypeS, mItemType)
+		}
+		typeable.SetType(&MapType{Items: mItemTypeS})
+	case *yamlmeta.ArrayItem:
+		typeable.SetType(&ArrayItemType{ValueType: a, Position: a.Position})
+		typeableValue, ok := typedItem.Value.(yamlmeta.Typeable)
+		if ok {
+			a.AssignTypeTo(typeableValue)
+		}
+	case *yamlmeta.MapItem:
+		typeable.SetType(&MapItemType{Key: typedItem.Key, ValueType: a, Position: a.Position})
+		typeableValue, ok := typedItem.Value.(yamlmeta.Typeable)
+		if ok {
+			a.AssignTypeTo(typeableValue)
+		}
+	}
+
+	return
+}
+
```
Random thought: what if the "zero value" of `Type` _is_ `AnyType`? That's another way of satisfying the "all Nodes have a Type" invariant...

Could be:
- all nodes come out of a constructor function
- that all nodes start life _with_ a valid for `Type`
- and that value is `AnyType`

...or...
- move all access to `Type` an accessor `GetType()`
- define that accessor to return `AnyType` when the underlying field is `nil`

Seems like it would make it easier to retain the invariant _and_ would obviate this whole function.

gcheadle-vmware replied:
`if node.Type == nil {return AnyType{}}`

pivotaljohn commented on pkg/schema/type.go:
```diff
@@ -117,6 +129,9 @@ func (m ScalarType) String() string {
 	}
 }
+func (a AnyType) String() string {
+	return "any type"
```
```suggestion
	return "any"
```

Based on, [Schemas — Part 3: Supported Types](https://hackmd.io/pODV3wzbT56MbQTxbQOOKQ?view#Part-3-Supported-Types) there exists _some_ semblance of a "convention" for this type namespace.

Until we GA the feature, this namespace (and frankly all others) are open to debate ... and then it becomes part of the published "API".

pivotaljohn commented on pkg/schema/type.go:
```diff
@@ -13,13 +13,16 @@ import (
 const (
 	AnnotationSchemaNullable structmeta.AnnotationName = "schema/nullable"
+	AnnotationSchemaType     structmeta.AnnotationName = "schema/type"
+	SchemaTypeAny            string                    = "any"
```
```suggestion
	TypeAnnotationKwargAny   string                    = "any"
```

Prior art:
https://github.com/vmware-tanzu/carvel-ytt/blob/b4a5e57e0b7b6d2d27edd857f8c8d3907172133b/pkg/yttlibrary/overlay/match_annotation_expects_kwarg.go#L16-L21

pivotaljohn commented on pkg/schema/type.go:
```diff
@@ -13,13 +13,16 @@ import (
 
 const (
 	AnnotationSchemaNullable structmeta.AnnotationName = "schema/nullable"
+	AnnotationSchemaType     structmeta.AnnotationName = "schema/type"
```
```suggestion
	AnnotationNullable structmeta.AnnotationName = "schema/nullable"
	AnnotationType     structmeta.AnnotationName = "schema/type"
```

Prior art:
https://github.com/vmware-tanzu/carvel-ytt/blob/b4a5e57e0b7b6d2d27edd857f8c8d3907172133b/pkg/yttlibrary/overlay/annotations.go#L14-L26

Also, subtle: when thinking about `type`s and `const`s, (especially exported ones), do include the package name in your consideration. Here, this means there's no need to (again) qualify the constant name. 👍🏼

pivotaljohn commented on pkg/schema/error.go:
```diff
@@ -137,6 +144,23 @@ func (t unexpectedKeyError) Error() string {
 	return msg
 }
 
+type invalidSchemaTypeAnnotationError struct {
```
```suggestion
type invalidTypeAnnotationError struct {
```

For brevity, can be shortened given that we're in the `schema` package.

pivotaljohn replied:
More fundamental question: whether we need a new error type here.

Is there a significant difference between an `invalidSchemaError` and a `invalidTypeAnnotationError`?  I personally haven't thought deeply about the `schema` package's errors.

I _think_:
- keep the number of `error` types to as few as possible (managing complexity)
- without losing meaningful distinctions (e.g. different data captured in each situation; different error message format desired; potentially different handling by callers...)

pivotaljohn commented on pkg/schema/type.go:
```diff
@@ -145,6 +160,9 @@ func (t *MapItemType) CheckType(node yamlmeta.TypeWithValues) (chk yamlmeta.Type
 		// A Map must've yielded a non-MapItem which is not valid YAML
 		panic(fmt.Sprintf("MapItem type check was called on a non-MapItem: %#v", node))
 	}
+	if _, anyType := t.ValueType.(AnyType); anyType {
+		return
+	}
```
Before the `AnyType` came along, we made "null" a special case (i.e. wrote explicit checks for null like the `if` just below here.

One consequence of that design choice is that it forced checking the _Value_ of a MapItem to be split:
- _sometimes_ occur in `MapItem.Check()` and,
-  _other times_ occur in the `{MapType|ArrayType|ScalarType}.CheckType()`

If we implement `@schema/nullable` as a `Type` then all the checks can occur in that second spot.
This leave the `MapItem.Check()` can be a simple structure check like it is in `ArrayItemType.CheckType()`.

So, while this wasn't a design choice made within _this_ story, we're forced to make matters worse... and this starts a downward spiral unless we jump off, right here.

Nothing is sacred and we're _almost always_ better informed today than we were last go around: let's make the `NullType` so that there's no need to do _any_ explicit check here.

pivotaljohn reviewed (changes requested):
Apologies for the delay in this review, @gcheadle-vmware.

Taking a full step back, `@schema/type any=True` is a profound add. It isn't just that a new annotation got added (and so, now there are two), or that it's the first schema annotation with kwargs...
... it critically highlights where certain parts of our design are wanting for a little more future proofing.

When navigating, sometimes we discover that what we _genuinely_ thought was a good course turns out to include trade-offs we — it turns out — don't want to make.

This is true for this story.

Here's the summary:

1. we didn't realize `@schema/nullable` as a `NullType`. That move resulted in unavoidable accidental complexity for this story. You stemmed the bleeding, but this is only going to get worse if we leave it, as is. (details in `schema/type.go`)
2. you clearly made a yeoman's effort in wrangling all the parts that reach from a given `Type` down through to the `@schema/type` annotation. Given that we're going to only expand our footprint, in this part of the code, we'd have more approachable code, but more extensible if we teased apart all the things into layers. I seeded that thinking with some pseudo code and a diagram. (see `schema/schema.go` for more).
3. the rest are some suggestions for nips and tucks, here and there.

gcheadle-vmware committed:
Add NullType to Type interface

- #@schema/nullable on an mapitem or array item has a type of NullType
- add schema/annotations.go to organize nullable and type annonations
- remove previous nullable checks that did not use the NullType

gcheadle-vmware committed:
WIP - add annotation tests to drive out refactor

DennisDenuto committed:
WIP - separate story

Signed-off-by: Garrett Cheadle <gcheadle@vmware.com>

gcheadle-vmware committed:
Generalize annotation processing in Schema
Signed-off-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Refactor:

- Remove 'dead code' from NullType
- Reduce responsibility of NullType to be a wrapper for the underlying
'ValueType' and during check to allow null values, otherwise, again,
delegate to the underlying 'ValueType' it is wrapping

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Refactor:

- Use a sort function to 'prioritize' AnyTypes over NullTypes when
deciding which 'Type' Based on a node's annotation to use

Authored-by: Dennis Leon <leonde@vmware.com>

gcheadle-vmware committed:
Refactor:

- Remove bool from NullableAnnotation struct
- Remove invalidSchemaAnnotationError, replace with just
  invalidSchemaError
- Clean up comments

Authored-by: Dennis Leon <leonde@vmware.com>

gcheadle-vmware committed:
Refactor:

- Remove bool from NullableAnnotation struct
- Remove invalidSchemaAnnotationError, replace with just
  invalidSchemaError
- Clean up comments

DennisDenuto committed:
Refactor:

- Make processNullableAnnotation && processTypeAnnotations 'similar'
  - processNullableAnnotation now uses a constructor to create a NullableAnnotation
  - a generic algorithm is used to process annotations on a node

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
UX: return error if schema/nullable has a kwarg

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Refactor:

- Make it clearer that when creating a map and arrayitem Type, that
nodes annotations take precedence over the nodes value
- Attempt to make functions/fields that I don't think need to be public ->
private. Mainly around the Annotation struct which is used internally as
a mapping from yaml annotations -> Types

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Add some context around skipped schema tests

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Refactor:

- Stop recursively creating a graph of AnyTypes (that mirrors the schema
yaml). Instead, during 'Check()', if the node's Type is nil,
assume it is an AnyType. Meaning, there isn't anything to check.
- Remove file checked in by accident

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Refactor:

- Make it clear that Node is a sealed interface. And why it is the case.

A sealed interface: can only be discussed in the
context of having multiple packages. A sealed interface is an interface
with unexported methods. This means users outside the package is unable
to create types that fulfil the interface.

In YTT's context, a Node must only ever have a type that maps
directly to a valid YAML type. Thus having an exhaustive list of
concrete types achieves that.

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Refactor:

- Return nil instead of empty ann for new annotation error case
- Stop using private member on Type annonation

Signed-off-by: Garrett Cheadle <gcheadle@vmware.com>

pivotaljohn reviewed (approved):
Yeomen's work, folks! Not only do we have support for the "any" type, but extended the type system to include a new "breed" — meta types that combine well with the existing primitive types.

🙏🏼
